pkg/che/tokencache.go

```diff
@@ -0,0 +1,156 @@
+package che
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	crtcfg "github.com/codeready-toolchain/member-operator/pkg/configuration"
+	"github.com/codeready-toolchain/member-operator/pkg/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/pkg/errors"
+)
+
+const tokenPath = "auth/realms/che/protocol/openid-connect/token"
+
+// tokenCache manages retrieving, caching and renewing the authentication token required for invoking Che APIs
+type tokenCache struct {
+	sync.RWMutex
+	httpClient *http.Client
+	token      *TokenSet
+}
+
+func newTokenCache() *tokenCache {
+	return &tokenCache{
+		httpClient: newHTTPClient(),
+	}
+}
+
+// getToken returns the token needed to use Che user APIs, or an error if there was a problem getting the token
+func (tc *tokenCache) getToken(cl client.Client, cfg *crtcfg.Config) (TokenSet, error) {
+	tc.RLock()
+	// use the cached credentials if they are still valid
+	if !tokenExpired(tc.token) {
+		defer tc.RUnlock()
+		log.Info("Reusing token")
+		return *tc.token, nil
+	}
+
+	tc.RUnlock()
+	defer tc.Unlock()
+	tc.Lock()
+	// get the credentials
+	user, pass := cfg.GetCheAdminUsername(), cfg.GetCheAdminPassword()
+	if user == "" || pass == "" {
+		return TokenSet{}, fmt.Errorf("the che admin username and/or password are not configured")
+	}
+
+	// get keycloak URL
+	cheKeycloakURL, err := getRouteURL(cl, cfg.GetCheNamespace(), cfg.GetCheKeycloakRouteName())
+	if err != nil {
+		return TokenSet{}, err
+	}
+	log.Info("Che Keycloak Route", "URL", cheKeycloakURL)
+
+	// get che user token
+	if err = tc.obtainNewToken(cheKeycloakURL+tokenPath, user, pass); err != nil {
+		return TokenSet{}, err
+	}
+
+	return *tc.token, nil
```

Comment:
I would rather move it to separated method so the read and write locks are more separated, but it's just my point of view ;-)

Reply:
Good suggestion, let me try splitting it up

Reply:
Done in 89b1e27e60e79ef25a5ba269afd473d55c6756eb